traversal/procedure/ProcedureEdge.java

```diff
@@ -504,19 +516,23 @@ public boolean isClosure(GraphManager graphMgr, Vertex<?, ?> fromVertex, Vertex<
                     }
 
                     @Override
-                    public FunctionalIterator<TypeVertex> branch(
-                            GraphManager graphMgr, Vertex<?, ?> fromVertex, GraphTraversal.Thing.Parameters params) {
+                    public Seekable<? extends Vertex<?, ?>, Order.Asc> branch(
+                            GraphManager graphMgr, Vertex<?, ?> fromVertex, Traversal.Parameters params) {
                         assert fromVertex.isType();
-                        FunctionalIterator<TypeVertex> iter;
+                        Seekable<TypeVertex, Order.Asc> iter;
                         TypeVertex type = fromVertex.asType();
                         if (!isTransitive) iter = type.ins().edge(SUB).from();
-                        else iter = tree(type, t -> t.ins().edge(SUB).from());
+                        else {
+                            TreeSet<TypeVertex> subtypes = new TreeSet<>();
+                            tree(type, t -> t.ins().edge(SUB).from()).forEachRemaining(subtypes::add);
+                            iter = iterateSorted(ASC, subtypes);
```

Comment:
note: this could be the souce of the performance regression
this would be unavoidable though!